azurelinuxagent/ga/collect_logs.py

```diff
@@ -0,0 +1,225 @@
+
+    def daemon(self, init_data=False):
+        try:
+            if not self._log_collection_allowed():
+                return
+
+            if init_data:
+                self.init_protocols()
+
+            while not self.stopped():
+                try:
+                    for op in self._periodic_operations: # pylint: disable=C0103
+                        op.run()
+                except Exception as e: # pylint: disable=C0103
+                    logger.error("An error occurred in the log collection thread main loop; "
+                                 "will skip the current iteration.\n{0}", ustr(e))
+                finally:
+                    PeriodicOperation.sleep_until_next_operation(self._periodic_operations)
+        except Exception as e: # pylint: disable=C0103
+            logger.error("An error occurred in the log collection thread; will exit the thread.\n{0}", ustr(e))
+
+    def collect_and_send_logs(self):
+        if self._collect_logs():
+            self._send_logs()
+
+    @staticmethod
+    def _get_resource_limits():
+        # Define CPU limit (as percentage of CPU time) and memory limit (absolute value in megabytes).
+        cpu_limit = "5%"
+        memory_limit = "30M"  # K for kb, M for mb
+        return cpu_limit, memory_limit
+
+    @staticmethod
+    def _collect_logs():
+        logger.info("Starting log collection...")
+
+        # Invoke the command line tool in the agent to collect logs, with resource limits on CPU and memory (RAM).
+        scope_name = "collect-logs-{0}.scope".format(ustr(int(time.time() * 1000000)))
+        systemd_cmd = ["systemd-run", "--unit={0}".format(scope_name), "--scope"]
+
+        # More info on resource limits properties in systemd here:
+        # https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/7/html/resource_management_guide/sec-modifying_control_groups
+        cpu_limit, memory_limit = CollectLogsHandler._get_resource_limits()
+        resource_limits = ["--property=CPUAccounting=1", "--property=CPUQuota={0}".format(cpu_limit),
+                           "--property=MemoryAccounting=1", "--property=MemoryLimit={0}".format(memory_limit)]
+
+        # The log tool is invoked from the current agent's egg with the command line option
+        collect_logs_cmd = [get_python_cmd(), "-u", sys.argv[0], "-collect-logs"]
+        final_command = systemd_cmd + resource_limits + collect_logs_cmd
+        start_time = datetime.datetime.utcnow()
+        try:
+            shellutil.run_command(final_command, log_error=True)
+            duration = elapsed_milliseconds(start_time)
+            archive_size = os.path.getsize(COMPRESSED_ARCHIVE_PATH)
+
+            msg = "Successfully collected logs. Archive size: {0} b, elapsed time: {1} ms.".format(archive_size,
+                                                                                                   duration)
+            logger.info(msg)
+            add_event(
```

Comment:
NIT: maybe add the `add_event` in `finally` clause with `msg and is_success` parameterized to make it look a bit cleaner and less redundant?

Reply:
Thanks, looks much cleaner.